src/ReinforcementLearningCore/src/core/stages.jl

```diff
@@ -20,3 +20,8 @@ Base.push!(p::AbstractPolicy, ::AbstractStage, ::AbstractEnv) = nothing
 Base.push!(p::AbstractPolicy, ::AbstractStage, ::AbstractEnv, ::Symbol) = nothing
 
 RLBase.optimise!(::AbstractPolicy) = nothing
+
+RLBase.optimise!(policy, ::PreEpisodeStage) = nothing
+RLBase.optimise!(policy::AbstractPolicy, ::PostEpisodeStage) = RLBase.optimise!(policy)
+RLBase.optimise!(policy, ::PreActStage) = nothing
+RLBase.optimise!(policy, ::PostActStage) = nothing
```

Comment:
This will introduce sneaky breaking changes to existing algorithms. I think by default all stages should be noop, or we keep it as is. If we change the default optimise stage in this PR, we must also fix the algorithms that optimise at each step.

Reply:
Also, I'm not sure why we would want to optimise at pre-act and pre-episode. But it doesn't hurt to have the calls I guess

Reply:
Ok nevermind, pre-act may be useful

Reply:
I think all should be defaulted to noop. Then, `optimise!(agent,...)` should dispatch on `agent.policy` + sample + stage. This dispatch is what should be overridden by the algorithm.

Reply:
I also think that we should keep the default behavior to update after every step. Intuitively, this is also the most natural form of interaction. Adaption is done after every decision. 

Also, PreEpisode and PostEpisode should not be differentiated, as there is nothing added to infer an update for the policy.

Reply:
The idea is that you optimise on the full information provided by the policy + the policy `push!` which happens at each stage. As a result, every stage *potentially* has a different information set, which may or may not be the right one for triggering `optimise!`. But perhaps this is from an RL perspective nonsense, happy to adjust.